In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("E:\\DATA ANALYTICS ENGINEERING MASTERY COURSE\\PANDAS MATERIAL\\policedata.csv")

In [3]:
df.head(2)

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   

   driver_age driver_race violation_raw violation  search_conducted  \
0        20.0       White      Speeding  Speeding             False   
1        40.0       White      Speeding  Speeding             False   

  search_type stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN     Citation       False      0-15 Min               False  
1         NaN     Citation       False      0-15 Min               False

In [4]:
df.shape

(65535, 15)

In [5]:
df.size

983025

In [6]:
df.index

RangeIndex(start=0, stop=65535, step=1)

In [7]:
df.columns

Index(['stop_date', 'stop_time', 'country_name', 'driver_gender',
       'driver_age_raw', 'driver_age', 'driver_race', 'violation_raw',
       'violation', 'search_conducted', 'search_type', 'stop_outcome',
       'is_arrested', 'stop_duration', 'drugs_related_stop'],
      dtype='object')

In [8]:
df.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

In [9]:
df.isnull().sum()*100/len(df)

stop_date               0.000000
stop_time               0.000000
country_name          100.000000
driver_gender           6.196689
driver_age_raw          6.186007
driver_age              6.572061
driver_race             6.195163
violation_raw           6.195163
violation               6.195163
search_conducted        0.000000
search_type            96.217288
stop_outcome            6.195163
is_arrested             6.195163
stop_duration           6.195163
drugs_related_stop      0.000000
dtype: float64

### Since here country_name & search_type containing null greater than 20 to 30 % so we have to delete it these two columns..

In [10]:
df.drop(columns=['country_name','search_type'],inplace=True)

In [11]:
df.isnull().sum()*100/len(df)

stop_date             0.000000
stop_time             0.000000
driver_gender         6.196689
driver_age_raw        6.186007
driver_age            6.572061
driver_race           6.195163
violation_raw         6.195163
violation             6.195163
search_conducted      0.000000
stop_outcome          6.195163
is_arrested           6.195163
stop_duration         6.195163
drugs_related_stop    0.000000
dtype: float64

### Now we can impute missing values of remaining col by measure of central tendency but first we need to explore datatype of each columns then we can decide so..

In [12]:
df.dtypes

stop_date              object
stop_time              object
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

In [13]:
df.isnull().sum()*100/len(df)

stop_date             0.000000
stop_time             0.000000
driver_gender         6.196689
driver_age_raw        6.186007
driver_age            6.572061
driver_race           6.195163
violation_raw         6.195163
violation             6.195163
search_conducted      0.000000
stop_outcome          6.195163
is_arrested           6.195163
stop_duration         6.195163
drugs_related_stop    0.000000
dtype: float64

# First convert datatype of stop_date col from object to datetime

In [14]:
df['stop_date']=pd.to_datetime(df['stop_date'])

In [15]:
df.dtypes

stop_date             datetime64[ns]
stop_time                     object
driver_gender                 object
driver_age_raw               float64
driver_age                   float64
driver_race                   object
violation_raw                 object
violation                     object
search_conducted                bool
stop_outcome                  object
is_arrested                   object
stop_duration                 object
drugs_related_stop              bool
dtype: object

#  We Can impute the missing values of 'Gender' col by mode of this col becoz it is Categorical datatype

In [16]:
df['driver_gender']=df.fillna(df['driver_gender'].mode())

In [17]:
df.isnull().sum()*100/len(df)

stop_date             0.000000
stop_time             0.000000
driver_gender         0.000000
driver_age_raw        6.186007
driver_age            6.572061
driver_race           6.195163
violation_raw         6.195163
violation             6.195163
search_conducted      0.000000
stop_outcome          6.195163
is_arrested           6.195163
stop_duration         6.195163
drugs_related_stop    0.000000
dtype: float64

In [18]:
df.dtypes

stop_date             datetime64[ns]
stop_time                     object
driver_gender                 object
driver_age_raw               float64
driver_age                   float64
driver_race                   object
violation_raw                 object
violation                     object
search_conducted                bool
stop_outcome                  object
is_arrested                   object
stop_duration                 object
drugs_related_stop              bool
dtype: object

In [19]:
df['driver_gender'].value_counts()

2012-01-11    69
2012-02-28    65
2007-05-16    62
2012-01-02    61
2007-01-10    60
              ..
2005-08-28     1
2005-04-01     1
2005-12-25     1
2005-09-03     1
2005-08-07     1
Name: driver_gender, Length: 2651, dtype: int64

# Now as remainig null are ver less so  we can drop also..

In [20]:
df.dropna(inplace=True)

# Now as We can see our dataset is fully cleaned..

In [21]:
df.isnull().sum()*100/len(df)

stop_date             0.0
stop_time             0.0
driver_gender         0.0
driver_age_raw        0.0
driver_age            0.0
driver_race           0.0
violation_raw         0.0
violation             0.0
search_conducted      0.0
stop_outcome          0.0
is_arrested           0.0
stop_duration         0.0
drugs_related_stop    0.0
dtype: float64

# Check Duplicated Records in dataset ?

In [22]:
df[df.duplicated()]

stop_date stop_time        driver_gender  driver_age_raw  driver_age  \
16    2005-07-19      0:30  2005-07-19 00:00:00          1982.0        23.0   
104   2005-10-03     13:26  2005-10-03 00:00:00          1975.0        30.0   
117   2005-10-04      0:36  2005-10-04 00:00:00          1961.0        44.0   
149   2005-10-05     10:00  2005-10-05 00:00:00          1981.0        24.0   
239   2005-10-08     12:30  2005-10-08 00:00:00          1978.0        27.0   
...          ...       ...                  ...             ...         ...   
62244 2012-07-26      9:45  2012-07-26 00:00:00          1953.0        59.0   
62474 2012-08-04     16:38  2012-08-04 00:00:00          1976.0        36.0   
63849 2012-09-21     10:14  2012-09-21 00:00:00          1979.0        33.0   
64323 2012-10-11      6:10  2012-10-11 00:00:00          1952.0        60.0   
65407 2012-11-30     22:16  2012-11-30 00:00:00          1990.0        22.0   

      driver_race                   violation_raw            violation  \
16          White                        Speeding             Speeding   
104         Black                        Speeding             Speeding   
117         White  Special Detail/Directed Patrol                Other   
149         White                        Speeding             Speeding   
239         White  Equipment/Inspection Violation            Equipment   
...           ...                             ...                  ...   
62244       White                        Speeding             Speeding   
62474       Black         Other Traffic Violation     Moving violation   
63849       White                        Speeding             Speeding   
64323       White          Registration Violation  Registration/plates   
65407       White                        Speeding             Speeding   

       search_conducted   stop_outcome is_arrested stop_duration  \
16                False       Citation       False      0-15 Min   
104               False  Arrest Driver        True       30+ Min   
117               False  Arrest Driver        True       30+ Min   
149               False       Citation       False      0-15 Min   
239               False       Citation       False      0-15 Min   
...                 ...            ...         ...           ...   
62244             False       Citation       False      0-15 Min   
62474             False       Citation       False      0-15 Min   
63849             False       Citation       False     16-30 Min   
64323             False       Citation       False      0-15 Min   
65407             False       Citation       False      0-15 Min   

       drugs_related_stop  
16                  False  
104                 False  
117                 False  
149                 False  
239                 False  
...                   ...  
62244               False  
62474               False  
63849               False  
64323               False  
65407               False  

[192 rows x 13 columns]

# So there are totally 192 records are duplicated so we can remove them ..

In [23]:
df.drop_duplicates(inplace=True)

# So Now if we check then we will get no duplicates records ?

In [24]:
df[df.duplicated()]

Empty DataFrame
Columns: [stop_date, stop_time, driver_gender, driver_age_raw, driver_age, driver_race, violation_raw, violation, search_conducted, stop_outcome, is_arrested, stop_duration, drugs_related_stop]
Index: []

In [25]:
df

stop_date stop_time        driver_gender  driver_age_raw  driver_age  \
0     2005-01-02      1:55  2005-01-02 00:00:00          1985.0        20.0   
1     2005-01-18      8:15  2005-01-18 00:00:00          1965.0        40.0   
2     2005-01-23     23:15  2005-01-23 00:00:00          1972.0        33.0   
3     2005-02-20     17:15  2005-02-20 00:00:00          1986.0        19.0   
4     2005-03-14     10:00  2005-03-14 00:00:00          1984.0        21.0   
...          ...       ...                  ...             ...         ...   
65529 2012-12-06     16:00  2012-12-06 00:00:00          1994.0        18.0   
65530 2012-12-06     17:54  2012-12-06 00:00:00          1987.0        25.0   
65531 2012-12-06     22:22  2012-12-06 00:00:00          1954.0        58.0   
65532 2012-12-06     23:20  2012-12-06 00:00:00          1985.0        27.0   
65534 2012-12-07      0:30  2012-12-07 00:00:00          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65529       White                        Speeding   Speeding   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65534       White                        Speeding   Speeding   

       search_conducted   stop_outcome is_arrested stop_duration  \
0                 False       Citation       False      0-15 Min   
1                 False       Citation       False      0-15 Min   
2                 False       Citation       False      0-15 Min   
3                 False  Arrest Driver        True     16-30 Min   
4                 False       Citation       False      0-15 Min   
...                 ...            ...         ...           ...   
65529             False       Citation       False      0-15 Min   
65530             False       Citation       False      0-15 Min   
65531             False        Warning       False      0-15 Min   
65532             False       Citation       False      0-15 Min   
65534             False       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65529               False  
65530               False  
65531               False  
65532               False  
65534               False  

[61031 rows x 13 columns]

# Now after deleting two col and deleting duplicated rows we left with only 61031 rows and 13 col with full cleaned dataset

In [26]:
df

stop_date stop_time        driver_gender  driver_age_raw  driver_age  \
0     2005-01-02      1:55  2005-01-02 00:00:00          1985.0        20.0   
1     2005-01-18      8:15  2005-01-18 00:00:00          1965.0        40.0   
2     2005-01-23     23:15  2005-01-23 00:00:00          1972.0        33.0   
3     2005-02-20     17:15  2005-02-20 00:00:00          1986.0        19.0   
4     2005-03-14     10:00  2005-03-14 00:00:00          1984.0        21.0   
...          ...       ...                  ...             ...         ...   
65529 2012-12-06     16:00  2012-12-06 00:00:00          1994.0        18.0   
65530 2012-12-06     17:54  2012-12-06 00:00:00          1987.0        25.0   
65531 2012-12-06     22:22  2012-12-06 00:00:00          1954.0        58.0   
65532 2012-12-06     23:20  2012-12-06 00:00:00          1985.0        27.0   
65534 2012-12-07      0:30  2012-12-07 00:00:00          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65529       White                        Speeding   Speeding   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65534       White                        Speeding   Speeding   

       search_conducted   stop_outcome is_arrested stop_duration  \
0                 False       Citation       False      0-15 Min   
1                 False       Citation       False      0-15 Min   
2                 False       Citation       False      0-15 Min   
3                 False  Arrest Driver        True     16-30 Min   
4                 False       Citation       False      0-15 Min   
...                 ...            ...         ...           ...   
65529             False       Citation       False      0-15 Min   
65530             False       Citation       False      0-15 Min   
65531             False        Warning       False      0-15 Min   
65532             False       Citation       False      0-15 Min   
65534             False       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65529               False  
65530               False  
65531               False  
65532               False  
65534               False  

[61031 rows x 13 columns]

In [29]:
df.drop(columns='driver_gender')

stop_date stop_time  driver_age_raw  driver_age driver_race  \
0     2005-01-02      1:55          1985.0        20.0       White   
1     2005-01-18      8:15          1965.0        40.0       White   
2     2005-01-23     23:15          1972.0        33.0       White   
3     2005-02-20     17:15          1986.0        19.0       White   
4     2005-03-14     10:00          1984.0        21.0       White   
...          ...       ...             ...         ...         ...   
65529 2012-12-06     16:00          1994.0        18.0       White   
65530 2012-12-06     17:54          1987.0        25.0       White   
65531 2012-12-06     22:22          1954.0        58.0       White   
65532 2012-12-06     23:20          1985.0        27.0       Black   
65534 2012-12-07      0:30          1985.0        27.0       White   

                        violation_raw  violation  search_conducted  \
0                            Speeding   Speeding             False   
1                            Speeding   Speeding             False   
2                            Speeding   Speeding             False   
3                    Call for Service      Other             False   
4                            Speeding   Speeding             False   
...                               ...        ...               ...   
65529                        Speeding   Speeding             False   
65530                        Speeding   Speeding             False   
65531                        Speeding   Speeding             False   
65532  Equipment/Inspection Violation  Equipment             False   
65534                        Speeding   Speeding             False   

        stop_outcome is_arrested stop_duration  drugs_related_stop  
0           Citation       False      0-15 Min               False  
1           Citation       False      0-15 Min               False  
2           Citation       False      0-15 Min               False  
3      Arrest Driver        True     16-30 Min               False  
4           Citation       False      0-15 Min               False  
...              ...         ...           ...                 ...  
65529       Citation       False      0-15 Min               False  
65530       Citation       False      0-15 Min               False  
65531        Warning       False      0-15 Min               False  
65532       Citation       False      0-15 Min               False  
65534       Citation       False      0-15 Min               False  

[61031 rows x 12 columns]

# What is the mean stop duration ?

In [30]:
df.head(1)

stop_date stop_time        driver_gender  driver_age_raw  driver_age  \
0 2005-01-02      1:55  2005-01-02 00:00:00          1985.0        20.0   

  driver_race violation_raw violation  search_conducted stop_outcome  \
0       White      Speeding  Speeding             False     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False

In [56]:
df['stop_duration'].value_counts()

0-15 Min     47080
16-30 Min    11331
30+ Min       2619
2                1
Name: stop_duration, dtype: int64

In [63]:
df['stop_duration'].replace(['0-15 Min','16-30 Min','30+ Min'],['7.5','23','45'],inplace=True)

In [69]:
df['stop_duration'].value_counts()

7.5     47080
23.0    11331
45.0     2619
2.0         1
Name: stop_duration, dtype: int64

In [ ]:
# Now Convert It's value to int type

In [67]:
df['stop_duration']=pd.to_numeric(df['stop_duration'])

In [68]:
df.dtypes

stop_date             datetime64[ns]
stop_time                     object
driver_gender                 object
driver_age_raw               float64
driver_age                   float64
driver_race                   object
violation_raw                 object
violation                     object
search_conducted                bool
stop_outcome                  object
is_arrested                   object
stop_duration                float64
drugs_related_stop              bool
dtype: object